src/executioners/SpectralExecutionerLinearElastic.C

```diff
@@ -0,0 +1,402 @@
+    const FFTBufferBase<RankFourTensor> & elastic_tensor)
+{
+  auto & stress_buffer = getFFTBuffer<RankTwoTensor>("stress");
+  const auto & grid = epsilon_buffer.grid();
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = grid[2];
+
+  // Homogeneous initial state of strain
+  fillOutEpsilonBuffer(epsilon_buffer);
+
+  size_t index = 0;
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        stress_buffer.realSpace()[index] =
+            elastic_tensor.realSpace()[index] * epsilon_buffer.realSpace()[index];
+        index++;
+      }
+
+  return stress_buffer;
+}
+
+void
+SpectralExecutionerLinearElastic::advanceReciprocalEpsilon(
+    FFTBufferBase<RankTwoTensor> & epsilon_buffer,
+    const FFTBufferBase<RankTwoTensor> & stress_buffer,
+    const FFTBufferBase<RankFourTensor> & gamma_hat)
+{
+  const auto & grid = epsilon_buffer.grid();
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = (grid[2] >> 1) + 1;
+
+  size_t index = 0;
+
+  const auto & ivec = gamma_hat.kTable(0);
+  const auto & jvec = gamma_hat.kTable(1);
+  const auto & kvec = gamma_hat.kTable(2);
+
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        if (ivec[freq_x] == 0 && jvec[freq_y] == 0 && kvec[freq_z] == 0)
+          epsilon_buffer.reciprocalSpace()[index] = _initial_strain_tensor;
+        else
+          epsilon_buffer.reciprocalSpace()[index] -=
+              gamma_hat.reciprocalSpace()[index] * stress_buffer.reciprocalSpace()[index];
+
+        index++;
+      }
+}
+
+void
+SpectralExecutionerLinearElastic::updateRealSigma(
+    const FFTBufferBase<RankTwoTensor> & epsilon_buffer,
+    FFTBufferBase<RankTwoTensor> & stress_buffer,
+    const FFTBufferBase<RankFourTensor> & elastic_tensor)
+{
+  const auto & grid = epsilon_buffer.grid();
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = grid[2];
+
+  size_t index = 0;
+
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        stress_buffer.realSpace()[index] =
+            (elastic_tensor.realSpace()[index]) * epsilon_buffer.realSpace()[index];
+        index++;
+      }
+}
+
+void
+SpectralExecutionerLinearElastic::filloutElasticTensor(
+    const FFTBufferBase<Real> & ratio_buffer,
+    FFTBufferBase<Real> & index_buffer,
+    FFTBufferBase<RankFourTensor> & elastic_tensor_buffer)
+{
+  const auto & grid = elastic_tensor_buffer.grid();
+
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = grid[2];
+
+  size_t index = 0;
+
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        // Define elastic tensor
+        std::vector<Real> iso_const(2);
```

Comment:
avoid creating short lived temporary vectors like the plague. They entail heap allocations and are thus very slow.

Reply:
I.e. move it out of the loop and since the size is known at compile time we could also think about using arrays (and templating the fillFormVector routines).